reactivesocket-core/src/main/java/io/reactivesocket/Frame.java

```diff
@@ -15,234 +15,267 @@
+     * @return {@code true} if and only if the reference count became {@code 0} and this object has been deallocated
      */
-    public void wrap(final MutableDirectBuffer directBuffer, final int offset) {
-        this.directBuffer = directBuffer;
-        this.offset = offset;
+    @Override
+    public boolean release() {
+        if (content.release()) {
+            recycle();
+            return true;
+        }
+        return false;
     }
 
     /**
-     * Acquire a free Frame backed by given ByteBuffer
-     * 
-     * @param byteBuffer to wrap
-     * @return new {@link Frame}
+     * Decreases the reference count by the specified {@code decrement} and deallocates this object if the reference
+     * count reaches at {@code 0}.
+     *
+     * @return {@code true} if and only if the reference count became {@code 0} and this object has been deallocated
      */
-    public static Frame from(final ByteBuffer byteBuffer) {
-        return POOL.acquireFrame(byteBuffer);
+    @Override
+    public boolean release(int decrement) {
+        if (content.release(decrement)) {
+            recycle();
+            return true;
+        }
+        return false;
     }
 
     /**
-     * Acquire a free Frame and back with the given {@link DirectBuffer} starting at offset for length bytes
+     * Return {@link ByteBuffer} that is a {@link ByteBuffer#slice()} for the frame metadata
      *
-     * @param directBuffer to use as backing buffer
-     * @param offset of start of frame
-     * @param length of frame in bytes
-     * @return frame
+     * If no metadata is present, the ByteBuffer will have 0 capacity.
+     *
+     * @return ByteBuffer containing the content
      */
-    public static Frame from(final DirectBuffer directBuffer, final int offset, final int length) {
-        final Frame frame = POOL.acquireFrame((MutableDirectBuffer)directBuffer);
-        frame.offset = offset;
-        frame.length = length;
-
-        return frame;
+    public ByteBuffer getMetadata() {
+        final ByteBuf metadata = FrameHeaderFlyweight.sliceFrameMetadata(content);
+        if (metadata.readableBytes() > 0) {
+            final ByteBuffer buffer = ByteBuffer.allocateDirect(metadata.readableBytes());
+            metadata.readBytes(buffer);
+            buffer.flip();
+            return buffer;
+        } else {
+            return NULL_BYTEBUFFER;
+        }
     }
 
     /**
-     * Construct a new Frame from the given {@link MutableDirectBuffer}
+     * Return {@link ByteBuffer} that is a {@link ByteBuffer#slice()} for the frame data
      *
-     * NOTE: always allocates. Used for pooling.
+     * If no data is present, the ByteBuffer will have 0 capacity.
      *
-     * @param directBuffer to wrap
-     * @return new {@link Frame}
+     * @return ByteBuffer containing the data
      */
-    public static Frame allocate(final MutableDirectBuffer directBuffer) {
-        return new Frame(directBuffer);
+    public ByteBuffer getData() {
```

Comment:
The main drawback of exposing ByteBuf in Payload from my point of view is that consumers of the library will have to worry about reference counting -- it would be nice to be zero copy all the way through, but I don't think it's worth the impact to usability.